src/sail_json_backend/json.ml

```diff
@@ -98,6 +99,21 @@ let dequote qs =
 
 let string_of_arg = function E_aux (E_id id, _) -> "\"" ^ string_of_id id ^ "\"" | exp -> "exp " ^ string_of_exp exp
 
+(* Function to parse the register type from the register name *)
+let json_of_registers () =
+  let regs = Hashtbl.fold (fun name regtype accum -> (name, regtype) :: accum) registers [] in
+  let sorted_regs =
+    List.sort (fun (n1, t1) (n2, t2) -> match String.compare t1 t2 with 0 -> String.compare n1 n2 | c -> c) regs
+  in
+  let json_regs =
+    List.map
+      (fun (name, regtype) ->
+        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
+      )
+      sorted_regs
+  in
+  "[ " ^ String.concat "" json_regs ^ "\n]"
```

Comment:
Actually, this has a similar problem to one I just fixed.

The emitted JSON for this content ends like:
```
  {
    "name": "vlenb",
    "type": "xlenbits"
  },
]
```
...it has a comma after the last element, which the Python JSON module does not like to load.

I think there's an easy fix: remove the comma at the end of line 111, and add it inside the empty string on line 115.

Here's how to test for valid JSON:
```
python3 -c "import json; f=open('AbhinavMir.out'); j=json.load(f)"
```

However, you'll need my recent fix (or delete the dangling comma at the end of the "functions" content). A rebase to my JSON branch should pull in the fix.

Reply:
Shold I just remove the comma or add it in line 115 to make it merge easy with latest code?

Reply:
Can I request a suggestion here? When I try to do the current suggested fix or removing a `,` and adding it back in on a later line, it gives me a still invalid JSON.

```ocaml
let json_of_registers () =
  let regs = Hashtbl.fold (fun name regtype accum -> (name, regtype) :: accum) registers [] in
  let sorted_regs =
    List.sort (fun (n1, t1) (n2, t2) -> match String.compare t1 t2 with 0 -> String.compare n1 n2 | c -> c) regs
  in
  let json_regs =
    List.map
      (fun (name, regtype) ->
        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
      )
      sorted_regs
  in
  "[ " ^ String.concat "" json_regs ^ "\n]"
```

Reply:
This works for me:
```
   let json_regs =
     List.map
       (fun (name, regtype) ->
-        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
+        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  }"
       )
       sorted_regs
   in
-  "[ " ^ String.concat "" json_regs ^ "\n]"
+  "[ " ^ String.concat "," json_regs ^ "\n]"
```

With your current branch state, that doesn't include 7ad136c5b2ba1e9c1cf9fef911a20173b7394042, you need to edit the emitted JSON to remove the dangling comma after the "functions" content:
```
  {
    "name": "riscv_f32Div",
    "source": "{\n  extern_f32Div(rm, v1, v2);\n  (float_fflags[4 .. 0], float_result[31 .. 0])\n}"
  }, (* <-- remove this comma *)
  ]
}
(* EOF *)
```

Reply:
I just dealt with that but it seems like there's still an unexpected `,` being generated towards the end of functions in the JSON. Is this the problem you dealt with right now but isn't merged into the code I forked? Just making sure.